File: algorithms/1_easy/F/02094_finding_3_digit_even_numbers/backtrack_with_counting.py
class Solution:
    def backtrack(
        self,
        digits_count:  List[int],
        res: List[int],
        curr_num: int,
        curr_depth: int,
    ):
        if curr_depth == 3:
            if curr_num >= 100 and curr_num < 1000 and curr_num % 2 == 0:
                res.append(curr_num)

            return

        for i in range(10):
            if digits_count[i] > 0:
                digits_count[i] -= 1
                self.backtrack(digits_count, res, curr_num * 10 + i, curr_depth + 1);
                digits_count[i] += 1

    def findEvenNumbers(self, digits: List[int]) -> List[int]:
        digits_count = [0] * 10

        for digit in digits:
            digits_count[digit] += 1

        res = []

        self.backtrack(digits_count, res, 0, 0)

        return res

File: algorithms/1_easy/F/02094_finding_3_digit_even_numbers/backtrack_with_counting.rs
impl Solution {
    fn backtrack(
        digits_count: &mut [i32; 10],
        res: &mut Vec<i32>,
        curr_num: i32,
        curr_depth: i32,
    ) {
        if curr_depth == 3 {
            if curr_num >= 100 && curr_num < 1000 && curr_num % 2 == 0 {
                res.push(curr_num);
            }

            return;
        }

        for i in 0..10 {
            if digits_count[i] > 0 {
                digits_count[i] -= 1;
                Self::backtrack(digits_count, res, curr_num * 10 + (i as i32), curr_depth + 1);
                digits_count[i] += 1;
            }
        }
    }

    pub fn find_even_numbers(mut digits: Vec<i32>) -> Vec<i32> {
        let mut digits_count = [0_i32; 10];

        for digit in digits.into_iter() {
            digits_count[digit as usize] += 1;
        }

        let mut res: Vec<i32> = Vec::new();

        Self::backtrack(&mut digits_count, &mut res, 0, 0);

        res
    }
}
